src/cargo/ops/cargo_generate_lockfile.rs

```diff
@@ -161,7 +164,49 @@ pub fn update_lockfile(ws: &Workspace<'_>, opts: &UpdateOptions<'_>) -> CargoRes
     let print_change = |status: &str, msg: String, color: &Style| {
         opts.config.shell().status_with_color(status, msg, color)
     };
-    for (removed, added) in compare_dependency_graphs(&previous_resolve, &resolve) {
+    let mut unchanged_behind = 0;
+    for (removed, added, unchanged) in compare_dependency_graphs(&previous_resolve, &resolve) {
+        let highest_present = [added.iter().rev().next(), unchanged.iter().rev().next()]
+            .into_iter()
+            .flatten()
+            .max_by_key(|s| s.version());
+        let latest = if let Some(present) = highest_present.filter(|p| p.source_id().is_registry())
+        {
+            let query = crate::core::dependency::Dependency::parse(
+                present.name(),
+                None,
+                present.source_id(),
+            )?;
+            let possibilities = loop {
+                match registry.query_vec(&query, QueryKind::Exact) {
+                    std::task::Poll::Ready(res) => {
+                        break res?;
+                    }
+                    std::task::Poll::Pending => registry.block_until_ready()?,
+                }
+            };
+            let warn = style::WARN;
+            let present_version = present.version();
+            possibilities
+                .iter()
+                .map(|s| s.as_summary())
+                .filter(|s| {
+                    let s_version = s.version();
+                    s_version.pre.is_empty()
+                        // Only match pre-release if major.minor.patch are the same
+                        || (s_version.major == present_version.major
+                            && s_version.minor == present_version.minor
+                            && s_version.patch == present_version.patch)
+                })
+                .map(|s| s.version().clone())
+                .max()
+                .filter(|v| present.version() < v)
+                .map(|v| format!(" {warn}(latest: v{v}){warn:#}"))
```

Comment:
Style suggestion: personally I find this kind of code easier to follow if you combine closures for `map().filter().map()` and `filter().map()` into one each.

Reply:
For me, its the opposite.  Each one is serving a unique purpose and having them represented separately better shows that and is easier to refactor (e.g. I should turn the `filter` into a `take_while`).  I also find `filter` for booleans easier to read than requiring converting bools to `Option` in a `filter_map`, whether through an explicit `if` or  through `.then`.

I use `filter_map` more when I'm working with `Option`s or the steps are more coupled.